### PR TITLE
Riot armor and helmets (and similar gear) protect against more melee-based attacks (like RNG punches), bottles aren't near guaranteed knockdowns

### PR DESCRIPTION
**Pull request by necromanceranne**
### Riot Armor and Helmets (including SWAT armor and plate armor; partially includes Syndicate modsuits and other suits with a module)

TRAIT_SHOVE_KNOCKDOWN_BLOCKED is now called TRAIT_BRAWLING_KNOCKDOWN_BLOCKED. This is possessed by riot suits, SWAT suits and now **plate armor**. Not the chaplain variety, because fuck them I guess. (this was apparently deliberate so don't complain to me, okay?)

Anyone with this trait is now unable to be knocked down by;

- Pure RNG on blunt objects attacks to the chest (Probability is otherwise altered by melee armor already)
- Unarmed punches, both RNG and deterministic
- Bodies thrown at them. Instead, dealing stamina damage and causing them to become staggered.

A new trait called TRAIT_HEAD_INJURY_BLOCKED, which protects from various head injury related effects. This is possessed by riot helmets, SWAT helmets, plate helmets, hardhats and applied to MODsuits with armor boosters activated/the infiltrator suit while active.

The trait is also granted to anyone wearing a modsuit with the Head Protection module installed. This can be printed from round start, and comes pre-installed into all the engineering modsuits, security moduits, research modsuit and magnate modsuit. (I had to bump up the security modsuit complexity a bit to put it in iunno if we have a policy about that)

Anyone with this trait is protected against:

- Bottle smashes to the head.
- RNG Blunt force trauma to the head by blunt objects (Probability is otherwise altered by melee armor already)
- Partially protects against getting your spine snapped during a bad tackle (this used to be a check ONLY for the riot helmet or hardhat)

### Bottle Smash

The bottle smash's chance of a knockdown is based on the force and knockdown duration of the bottle, altered by relevant head melee armor. It's no longer basically guaranteed due to weird math that would determine whether or not you were knocked down only if your armor exceeded certain values or not.

### Misc Changes

Any instance that would check either for riot armor or riot helmets instead checks for the new traits.

## Why It's Good For The Game

People weren't particularly happy with the possibility of getting ownzoned by a naked assistant or random toolbox-wielding tider while in dedicated melee protection armor as a result of pure RNG jank. [There was a whole thread here about it](https://tgstation13.org/phpBB/viewtopic.php?f=83&t=35645) with regards to unarmed combat.

If you're in riot gear, you should feel like you can confidently combat improvised/blunt weapons at a cost of general protection from the more serious dangers to your wellbeing (AKA lead poisoning or heatstroke. Get it....it's bullets and lasers...)

For some reason, bottle smash knockdowns, despite using the values just...conveniently don't get blocked by armor except for extreme values? If at all? I felt like this was really weird for something with such a hefty and fairly powerful knockdown effect, particularly one you can lob at someones head at range. I remember, way back when I first started playing, that this was a feature that used to stun, and one I used a lot to get cheap kills. The amount of bullshit stuns I got on people with it still haunt me to this day. It hasn't improved in the current era, despite being a knockdown, simply because any knockdown still takes several seconds to exit. An arbitrarily low knockdown is still a several second one. It's time to bring this in line with similar equipment.

Rather than using flat values that are _clearly_ meant to be only applicable to riot armor the decade ago that the code was written, let's use traits instead. That's way better than relying on magic numbers that may become irrelevant when the code around it changes, which anyone touching that code may not even be aware existed.

Since we're using a trait, we can actually allow modsuits to occasionally fill defensive niches that aren't strictly armor values. Handy.

## Changelog

:cl:
balance: Nanotrasen, in direct response to the increasing danger posed by wannabe martial artists and rioters in the fringes of the Spinward Sector, have upgraded the impact dampeners found in their riot armor. Staff have also started to rediscover the value of medieval armor; it isn't particularly easy to topple a knight in a suit of plate with just your fists.
balance: Melee-focused armor is now more able to protect you from various RNG-based knockdowns, such as critical hits from punches (as well as the ones applied through the staggered status), shoves, critical hits with a blunt weapon to the chest, and body throws.
balance: Melee-focused helmets also protect you from head injuries, such as bottle smashes, accidentally hitting something dense during a tackle, and critical hits from a blunt weapon to the head.
balance: Bottlesmash knockdowns are less reliable in general.
add: A new module, the Safety-First Head Protection module, protects you from head trauma! Available in most modsuits expected to take hits to the head often. And from roundstart exofabricators.
/:cl:

### CODE REVIEW

**Comment by SomeoneYouProbablyKnow** [comment]
Body throw block? Daym. Time to invest in riot armour.

**Comment by LePiromano** [comment]
how bout giving this to miner armors? at least drake and hecksuit, no?

**Comment by Loafin34** [comment]
add this to nukiesuit, infilsuit, and make the bottle modifier on all "hard" material hats. stuff like welding helmets that are up, hard hats, secoff helmets (not the cap or beret tho) and bucket.

**Comment by necromanceranne** [comment]
I feel like if I'm adding knockdown proofing to nukies, it's really something to atomize because that changes so fucking much about how you approach dealing with them.

**Comment by necromanceranne** [comment]
> how bout giving this to miner armors? at least drake and hecksuit, they're melee-focused armor afterall, no?

Lol

**Comment by Xackii** [comment]
RnG is shit. Make RnG on stun baton or dislike.

**Comment by necromanceranne** [comment]
I mean, it has RNG on a throw impact.

**Comment by necromanceranne** [comment]
Hardhats now have the trait. I really should have added in to begin with but whatevs

**Comment by necromanceranne** [comment]
I added a method to make modsuits able to protect you from head injuries with a module. All the engineering suits, security suits, the research modsuit and the magnate modsuit have it pre-installed. It is also a roundstart printable module, so if you want it, you can go get it.

**Comment by breadcrombs** [comment]
having resistance to complete body throws kinda it op, maybe make it deal huge amounts of stam damage or have knockdown without the stun

**Comment by necromanceranne** [comment]
Hm, that's a reasonable alternative. Sure.

**Comment by moocowswag** [comment]
Body throws should really use a look and made to scale with/respect throwforce levels if they dont already

**Comment by necromanceranne** [comment]
Oh, they do. But it's a binary 'is it a harmful throw/is it not a harmful throw' check, rather than the scale/speed of the throw impacts whether the throw will cause a knockdown.

**Comment by DrAmazing343** [comment]
Another banger necromanceranne PR. Good on ya for following up on the community sentiments about the melee stuff; I'll always support solid balancing stuff that doesn't ruin rule of cool, and this is aptly fair for armor types that are dedicated to anti-melee/unarmed.

**Comment by Xackii** [comment]
brawling stun baton when?

**Comment by Dorsisdwarf** [comment]
Always good to see dedicated anti-greytide kit get benefits against the mob that arent just generic power buffs

**Comment by bigfatbananacyclops** [comment]
this will make riot armor a meta regardless of the rng feature.

File: code/_globalvars/traits/_traits.dm [hunk 1]
```diff
@@ -397,7 +397,8 @@ GLOBAL_LIST_INIT(traits_by_type, list(
 		"TRAIT_SHAVED" = TRAIT_SHAVED,
 		"TRAIT_SHIFTY_EYES" = TRAIT_SHIFTY_EYES,
 		"TRAIT_SHOCKIMMUNE" = TRAIT_SHOCKIMMUNE,
-		"TRAIT_SHOVE_KNOCKDOWN_BLOCKED" = TRAIT_SHOVE_KNOCKDOWN_BLOCKED,
+		"TRAIT_BRAWLING_KNOCKDOWN_BLOCKED" = TRAIT_BRAWLING_KNOCKDOWN_BLOCKED,
+		"TRAIT_HEAD_INJURY_BLOCKED" = TRAIT_HEAD_INJURY_BLOCKED,
```
Comment (Ghommie) [inline code/_globalvars/traits/_traits.dm:401]: Alphabetical order, please.

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Comment by ExcessiveUseOfCobblestone** [comment]
> RnG is shit. Make RnG on stun baton or dislike.

There is nothing wrong with RNG, it attempts to equalize the nature of the pvp with the fact it's a RP game by allowing crew who have not spent the entire round with pvp in mind a way to overcome. Take a look at disarming, if you get caught without knockdown/stun gear (because youre not playing a combat-oriented job and dont want to game) versus someone who has made it their sole purpose in the round to build up for combat AND they know not to be near a wall like most veteran players do, you are NOT knocking them down. 

Be mindful anything crew has an antag with hands also has the potential to access, either to use themselves or to discard.

That said, im ok with the PR but we really should be looking at understanding that if crew gets punished by not having gear because they lose their equalizers, the only real option is to encourage rushing a baton or other gamer equipment-of-choice. It's an extremely fickled balance. We need to make sure that at no point all equalizers can be blocked off simultaneously to the point theres nothing you can do unless youve preemptively built/stole something (which is what people will start doing).

File: code/_globalvars/traits/_traits.dm (hunk 1 above)
Comment (necromanceranne) [reply code/_globalvars/traits/_traits.dm:401]: aye, one sec

Comment (necromanceranne) [reply code/_globalvars/traits/_traits.dm]: done

**Comment by necromanceranne** [comment]
If the result is that folk are overutilizing the armor because people aren't really able to do anything to someone in it, then I'll walk back some of these changes/alter them in some way.

**Comment by deathrobotpunch** [comment]
> I added a method to make modsuits able to protect you from head injuries with a module. All the engineering suits, security suits, the research modsuit and the magnate modsuit have it pre-installed. It is also a roundstart printable module, so if you want it, you can go get it.

rocket glove power creep

**Comment by DrDiasyl** [comment]
> Just saying, this will make riot armor a meta preference, regardless of the rng. The only antags that get ballistics are nukies and traitor(only gets a makarov), wearing a bullet-proof wont make a difference. So everyone will be more inclined to wear riot suits, especially because it's effective against almost every form of antag(whom most rely on melee).
> 
> The important factors the riot armor should have is the immunity of being wall shoved, eliminating rng unarmed knockdown and reducing the effects of bleeding caused by punctures or sharp weapons. All exception of energy based swords.
> 
> Body slam is something that should be, and currently is(based on rng), blocked by shields and also the sec ripley. It should block bodyslams, with the consequence of the stamina damage you proposed.
> 
> Modsuits having this shove immunity, puncture and body-slam immunity is just gonna make them a fast-walking mech. Again, i think the bodyslam proof is a better feature for shields.
> 
> If you're going to add those sort of buffs(especially the body-slam)then the speed that a riot armor provides should be considered to balance it out. Because it is crazy good advantage. You might be able to slowdown/stam damage a riot officer but they can still shoot you, and they might be on drugs.

Speed debuffs are the worst kind of debuffs.
Being slower then anyone by even a little means too much in our combat.

**Comment by DrDiasyl** [comment]
This PR is great because it makes melee combat not always better than ranged combat.
Melee combat was better than ranged combat mostly due all of the stuns being in melee, this way when they're not a great option you will have to use guns in more situations.

**Comment by SinfulBliss** [comment]
> The bottle smash's chance of a knockdown is based on the force and knockdown duration of the bottle, altered by relevant head melee armor. It's no longer basically guaranteed due to weird math that would determine whether or not you were knocked down only if your armor exceeded certain values or not.
Tested this the other day - I haven't noticed any affect to the RNG based on the type of bottle. Big bottles of vodka and other large bottles have the same RNG as a beer bottle. 

The RNG is also too weak to make bottling worth it at all anymore, it feels like it's around a 20-25% chance for a knockdown even against someone completely naked with no head armor.

**Pull request merged**